packages/sources/tradingeconomics-test/src/endpoint/price-ws.ts

```diff
@@ -0,0 +1,104 @@
+import { WebSocketTransport } from '@chainlink/external-adapter-framework/transports'
+import { makeLogger, SingleNumberResultResponse } from '@chainlink/external-adapter-framework/util'
+import includes from '../config/includes.json'
+
+import { customSettings } from '../config'
+import { PriceEndpointParams } from '@chainlink/external-adapter-framework/adapter'
+
+const logger = makeLogger('Tradingeconomics WS')
+
+export type EndpointTypes = {
+  Request: {
+    Params: PriceEndpointParams
+  }
+  Response: SingleNumberResultResponse
+  CustomSettings: typeof customSettings
+  Provider: {
+    WsMessage: Message[]
+  }
+}
+
+type Includes = {
+  from: string
+  to: string
+  includes: IncludePair[]
+}
+
+export type IncludePair = {
+  from: string // From symbol
+  to: string // To symbol
+  adapters?: string[] // Array of adapters this applies to
+  inverse?: boolean // If the inverse should be calculated instead
+  tokens?: boolean // If the token addresses should be used instead
+}
+
+interface Message {
+  s: string
+  i: string
+  pch: number
+  nch: number
+  bid: number
+  ask: number
+  price: number
+  dt: number
+  state: string
+  type: string
+  dhigh: number
+  dlow: number
+  o: number
+  prev: number
+  topic: string
+}
+
+const withApiKey = (url: string, key: string, secret: string) => `${url}?client=${key}:${secret}`
+const getSubscription = (to: string) => ({ topic: 'subscribe', to })
+
+const baseFromIncludes = includes.reduce(
+  (basesMap: { [from: string]: string }, includesSet: Includes) => {
+    const { includes } = includesSet
+    for (const includePair of includes) {
+      basesMap[includePair.from] = includesSet.from
+    }
+    return basesMap
+  },
+  {},
+)
+
+export const wsTransport = new WebSocketTransport<EndpointTypes>({
+  url: (context) => {
+    const { API_CLIENT_KEY, API_CLIENT_SECRET, WS_API_ENDPOINT } = context.adapterConfig
+    return withApiKey(WS_API_ENDPOINT, API_CLIENT_KEY || '', API_CLIENT_SECRET || '')
+  },
+  handlers: {
+    message: (message) => {
+      if (Object.keys(message).length === 0) {
+        logger.debug('WS message is empty, skipping')
+        return []
+      }
+      return message.map((msg) => {
+        const base = baseFromIncludes[msg?.s] ?? msg?.s
+        const result = {
+          params: { base, quote: 'USD' },
```

Comment:
Is quote always "USD"? Seems like we should use the input param here

Reply:
thats how they do in v2:

```
 const base = baseFromIncludes[wsResponse?.s] ?? wsResponse?.s
        const validator = new Validator<endpoints.TInputParameters>(
          { id: 'tradingeconomics-ws-to-repsonse', data: { base, quote: 'USD' } },
          inputParameters,
          {},
          { shouldThrowError: false, overrides, includes },
        )
```
There is an issue with v2, because it matches a predefined list in `includes.json` with the default quote in `USD` and outputs the query depending on which one it matches. Can you take a look at v2 pls?

example of the `includes.json` file:
```
[
  {
    "from": "CAD",
    "to": "USD",
    "includes": [
      {
        "from": "USDCAD:CUR",
        "to": "",
        "inverse": true
      }
    ]
  },
  {
    "from": "CHF",
    "to": "USD",
    "includes": [
      {
        "from": "USDCHF:CUR",
        "to": "",
        "inverse": true
      }
    ]
  },
  ...
  ```

Reply:
I see. It's not ideal, but let's keep this so we don't break anything I guess :man_shrugging: